src/octoprint/plugins/pluginmanager/__init__.py

```diff
@@ -497,6 +511,38 @@ def condition():
             unless=lambda: refresh,
         )(view)()
 
+    @octoprint.plugin.BlueprintPlugin.route("/plugins/versions")
+    @Permissions.PLUGIN_PLUGINMANAGER_LIST.require(403)
+    def retrieve_plugin_list(self):
+        refresh = request.values.get("refresh", "false") in valid_boolean_trues
+        if refresh or not self._is_notices_cache_valid():
+            self._notices_available = self._refresh_notices()
+
+        def view():
+            return jsonify(**self._plugin_versions_response())
+
+        def etag():
+            import hashlib
+
+            hash = hashlib.sha1()
+
+            def hash_update(value):
+                value = value.encode("utf-8")
+                hash.update(value)
+
+            hash_update(repr(self._plugin_versions_response()))
```

Comment:
This will behave non-deterministically depending on the used python versions, because the order of items returned from a dict is not guaranteed to match insertion order in Python < 3.7.

Better:

``` python
hash_update(repr(list(map(lambda k, v: "{}={}".format(k, v), self._plugins_versions_response()))))
```

Alternatively you could also use `OrderedDict` for building the data structure.